components/common/Cargo.toml

```diff
@@ -54,6 +58,7 @@ tokio = { version = "*", features = ["full"] }
 toml = { version = "*", features = ["preserve_order"] }
 uuid = { version = "*", features = ["v4"] }
 valico = "*"
+url = { version = "2.5" }
```

Comment:
Actually, using full wild card dependencies is not considered a good practice eg. `crates.io` does not accept packages with wildcard dependencies anymore (for a while). For now that's not a problem for us per se, but it's a good idea that we start moving away from wildcard dependencies. 

That's a reason, I decided to start this in a small self contained crate and then we can go ahead with more crates step by step. 

One of the reasons why this is a bad practice is - we do not know which versions are being used and if some version gets yanked, we start seeing a failure, I believe that started happening recently with some really old crate that was yanked #9322

Reply:
That totally makes sense for crates that are built for external consumption. However our hab crates are top level and only support our own application. We don't expect external developers to build our components without a `Cargo.lock` file.

To the extent it is reasonable, we try to keep all of our dependencies as current as possible and manage our builds with the committed `Cargo.lock` file. If any dep gets yanked, we have a `cargo audit` job in our build to catch this and break the pipeline so we fix it.

The benefits of the wildcard for us have been that it lets dependabot or `cargo update` continually bump our deps. We then only pin or restrict where we know a later version would break us and require more work to fix than we could do ad hoc. 

I'm totally open to a different approach but wanted to clarify the reasoning here. Does it make sense?

Reply:
Dependabot would normally update patched versions  and sometimes minor versions.  Dependabot would just as well work with well defined versions. (I need to confirm it - but I am reasonably sure! :-) ). 

I think at-least having a strategy of well-defined major (and when it doesn't make sense - for example when major version is `0`) and minor versions is a good choice, and even streamlining those versions across crates (eg. if it is 2.5 in one crate, it will be the same version in another crate), slowly we should try to get there. 

What I dislike about wildcard versions is - it's not trivial to look up at `Cargo.toml` and find out which versions we are using. 

And when things break, we anyways have to Pin those versions, so why not start right away? 

Unlikely, but for whatever reasons if we decide to publish even some of our crates, it will be a big exercise then like the `clap` one right now, so while we can, it's a good idea to go ahead and have well defined versions. 

I believe aligning with best practices followed elsewhere in the ecosystem is a good long term goal. 

Makes sense?

Reply:
Lets discuss offline in standup.